apps/studio/src/components/ResourceSelector/useResourceStack.tsx

```diff
@@ -0,0 +1,209 @@
+
+      // A resource should not be able to move to within itself
+      if (existingResource.id === resourceItem.id) return true
+
+      // If a resource is not allowed to have children then it is a page-ish resource
+      // Thus, it can move to within any resource and no further checks are needed
+      if (!isAllowedToHaveChildren(existingResource.type)) return false
+
+      // A resource should not be able to move to its nested children
+      return (
+        nestedChildrenOfExistingResource.some(
+          (child) => child.id === resourceItem.id,
+        ) || false
+      )
+    },
+    [existingResource, nestedChildrenOfExistingResource],
+  )
+
+  const hasParentInStack = useMemo(
+    () =>
+      (resourceStack.length === 1 && !isResourceHighlighted) ||
+      resourceStack.length > 1,
+    [resourceStack.length, isResourceHighlighted],
+  )
+
+  const fullPermalink: string = useMemo(() => {
+    return resourceStack.map((resource) => resource.permalink).join("/")
+  }, [resourceStack])
+
+  const handleClickResourceItem = (
+    resourceItemWithAncestryStack: ResourceItemContent[],
+  ): void => {
+    const lastChild: ResourceItemContent | undefined =
+      lastResourceItemInAncestryStack(resourceItemWithAncestryStack)
+
+    // this should never happen. only added here to satisfy typescript
+    if (!lastChild) return
+
+    const isItemHighlighted = isResourceIdHighlighted(lastChild.id)
+    const canClickIntoItem =
+      lastChild.type === ResourceType.Folder ||
+      lastChild.type === ResourceType.Collection
+
+    if (isItemHighlighted && canClickIntoItem) {
+      setIsResourceHighlighted(false)
+      return
+    }
+
+    setResourceStack(resourceItemWithAncestryStack)
+    setIsResourceHighlighted(true)
+  }
+
+  const handleClickBackButton = useCallback(() => {
+    if (isResourceHighlighted) {
+      setIsResourceHighlighted(false)
+      removeFromStack(2)
+    } else {
+      removeFromStack(1)
+    }
+  }, [isResourceHighlighted, removeFromStack])
+
+  const pendingMovedItemAncestryStack: ResourceItemContent[] =
+    trpc.resource.getAncestryWithSelf
+      .useSuspenseQuery({
+        siteId: String(siteId),
+        resourceId: selectedResourceId,
+      })[0]
+      .map((resource) => ({ ...resource, resourceId: resource.id }))
+
+  useEffect(() => {
+    // If there is no selected resource, we don't need to update the stack
+    if (!selectedResourceId) return
+
+    // If the ancestry stack is empty, we don't need to update the stack
+    if (pendingMovedItemAncestryStack.length <= 0) return
+
+    // If the ancestry stack is the same as the current stack, we don't need to update the stack
+    if (
+      JSON.stringify(pendingMovedItemAncestryStack) ===
+      JSON.stringify(resourceStack)
+    ) {
+      return
+    }
+
+    setResourceStack(pendingMovedItemAncestryStack)
+  }, [])
+
+  useEffect(() => {
+    if (curResourceId) {
+      onChange(curResourceId)
+    }
+  }, [curResourceId])
```

Comment:
[you might not need an effect](https://react.dev/learn/you-might-not-need-an-effect)

you update the resource stack only when you click on the resource item. You might want to trigger `onChange` there (inside `handleClickResourceItem`).